### PR TITLE
Tazmen sniffer protocol layer

### PR DESCRIPTION
**Pull request by speakinghedge**
This adds the TaZmen sniffer protocol layer.

Its not completely clear which source of the header definition (wikipedia/networkchemistry.com@archive.org or the wireshark dissector) contains the 'truth' -
but this works with the implementation of TZSP used in my Mikrotik router.

this is based on a request in #1000 (but I found it useful for me too so ...)

### CODE REVIEW

File: scapy/contrib/tzsp.uts [hunk 1]
```diff
@@ -0,0 +1,503 @@
...
+
+tzsp_tag_padding = tzsp_tag_sensor_id.payload
+assert(tzsp_tag_padding.type == 0)
+
+tzsp_tag_end = tzsp_tag_padding.payload
+assert(tzsp_tag_end.type == 1)
+
+encapsulated_payload = tzsp_tag_end.payload
+encapsulated_ether_lyr = encapsulated_payload.getlayer(Ether)
+assert(encapsulated_ether_lyr.src == '00:03:03:03:03:03')
```
Comment (gpotter2) [inline scapy/contrib/tzsp.uts]: You could add a newline :)

File: scapy/contrib/tzsp.uts [hunk 2]
```diff
@@ -0,0 +1,503 @@
+from enum import test
```
Comment (gpotter2) [inline scapy/contrib/tzsp.uts]: Are you sure that this is needed ? What is it used for ?

File: scapy/contrib/tzsp.py [hunk 3]
```diff
@@ -0,0 +1,474 @@
...
+          - doesn't know the packet count tag (40 / 0x28)
+
+"""
+from scapy.contrib.avs import AVSWLANHeader
+from scapy.error import warning
+from scapy.fields import ByteField, ShortEnumField, IntField, FieldLenField
+from scapy.layers.dot11 import Packet, Dot11, PrismHeader
+from scapy.layers.l2 import Ether, \
+    struct, StrLenField, ByteEnumField, \
+    ShortField, Scapy_Exception
```
Comment (gpotter2) [inline scapy/contrib/tzsp.py]: Please import things from their correct file:
```
Scapy_Exception                        --> scapy.error
StrLenField, ByteEnumField, ShortField --> scapy.fields
struct                                 --> struct
```

File: scapy/contrib/tzsp.uts [hunk 4]
```diff
@@ -0,0 +1,503 @@
...
+
+#
+# execute test:
+# > test/run_tests -P "load_contrib('tzsp')" -t scapy/contrib/tzsp.uts
+#
+
++ syntax check
+= check if the import works
+
+from scapy.contrib.tzsp import *
```
Comment (p-l-) [inline scapy/contrib/tzsp.uts]: This is not needed as we run the tests with the appropriate option to load the contrib module.

File: scapy/contrib/tzsp.py (hunk 3 above)
Comment (speakinghedge) [reply scapy/contrib/tzsp.py]: thanks for the review. seems pycharm took a shortcut when adding those imports automatically - but at the end of the day its my duty...

File: scapy/contrib/tzsp.uts (hunk 2 above)
Comment (speakinghedge) [reply scapy/contrib/tzsp.uts]: not used at all - removed.

File: scapy/contrib/tzsp.uts (hunk 1 above)
Comment (speakinghedge) [reply scapy/contrib/tzsp.uts]: yes

File: scapy/contrib/tzsp.uts (hunk 4 above)
Comment (speakinghedge) [reply scapy/contrib/tzsp.uts]: but this gives a clean indication if something is wrong with the imported file - but I will remove it for the sake of consistency.

File: scapy/contrib/tzsp.py [hunk 5]
```diff
@@ -357,14 +358,10 @@ class TZSPTagContentionFree(_TZSPTag):
...
-    YES = 0x01
     NO = 0x00
+    YES = 0x01
 
-    CONTENTION_STATES = {
-        NO: 'no',
-        YES: 'yes',
-        range(2, 0xff): 'yes'
-    }
+    CONTENTION_STATES = {key: ('yes' if key != 0x00 else 'no') for key in range(0x00, 0x100)}
```
Comment (speakinghedge) [inline scapy/contrib/tzsp.py]: @p-l- 
@gpotter2 
is this okay or to ugly (in any terms)?

**Comment by speakinghedge** [comment]
plz call me back if you are done here so I can cleanup the commits

File: scapy/contrib/tzsp.py [hunk 6]
```diff
@@ -0,0 +1,493 @@
...
+
+
+class TZSPTagContentionFree(_TZSPTag):
+    """
+    packet received in contention free period
+    """
+    NO = 0x00
+    YES = 0x01
+
+    CONTENTION_STATES = {key: ('yes' if key != 0x00 else 'no') for key in range(0x00, 0x100)}
```
Comment (gpotter2) [inline scapy/contrib/tzsp.py]: This is extremely ugly. As you use the same thing several times, I suggest you to use something such as:
```
class YesNoByteField(ByteField):
    __slots__ = [“yes”]
    def __init__(self, name, default, yes=0, *args, **kargs):
        self.yes = yes
        ByteField.__init__(self, name, default, *args, **kargs)
    def i2repr(self, pkt, x):
        return “yes” if x == self.yes else “no”
```

Which you may want to add to `fields.py` for further usages, as quite common

Comment (speakinghedge) [reply scapy/contrib/tzsp.py]: good idea. as TZSP also needs a tri-state-representation (0 - no, 1 - yes, 2..ff - reserved) I added a slightly more flexible version than your proposed solution (ed19f9b).

Comment (gpotter2) [reply scapy/contrib/tzsp.py]: Good job !

**Comment by guedou** [comment]
I lost track of this PR. @gpotter2 is it ok to merge it ?

**Pull request merged**